specification/dartLangSpec.tex

```diff
@@ -2531,62 +2531,20 @@ \section{Classes}
 that name denotes the interface of the class.
 
 \LMHash{}%
-% The use of `concrete member' below may seem redundant, because a class
-% does not inherit abstract members from its superclass, but this
-% underscores the fact that even when an abstract declaration of $m$ is
-% declared in $C$, $C$ does not "have" $m$.
-A concrete class must fully implement its interface:
-\BlindDefineSymbol{C, I, m}%
-Let $C$ be a concrete class with interface $I$.
-Assume that $I$ has an accessible member signature $m$,
-and $C$ does not declare a member with the same name as $m$.
-It is a compile-time error if $C$ does not have
-a concrete accessible member with the same name as $m$,
-unless $C$ has a non-trivial \code{noSuchMethod}
-(\ref{theMethodNoSuchMethod}).
-It is a compile-time error if $C$ has
-a concrete accessible member with the same name as $m$,
-with a method signature $m'$ which is not a correct override of $m$
-(\ref{correctMemberOverrides}),
-unless that concrete member is a \code{noSuchMethod} forwarder
-(\ref{theMethodNoSuchMethod}).
-
-\commentary{%
-In particular, it is an error for a class to be concrete even if it inherits
-a member implementation for every member signature in its interface,
-unless each of them has parameters and types such that they satisfy
-the corresponding member signature.
-But when there is a non-trivial \code{noSuchMethod} it is allowed
-to leave some members unimplemented,
-and it is allowed to have a \code{noSuchMethod} forwarder which does not
-satisfy the class interface
-(in which case it will be overridden by another \code{noSuchMethod} forwarder).%
-}
-
-\LMHash{}%
-Assume that $C$ has a concrete accessible member with the same name as $m$,
-with a method signature $m'$ which is a correct override of $m$.
-For each parameter $p$ of $m'$ where \COVARIANT{} is present,
-it is a compile-time error if there exists
-a direct or indirect superinterface of $C$ which has
-an accessible method signature $m''$ with the same name as $m$,
-such that $m''$ has a parameter $p''$ that corresponds to $p$
-(\ref{covariantParameters}),
-unless the type of $p$ is a subtype or a supertype of the type of $p''$.
-
-\commentary{%
-This ensures that an inherited method satisfies the same constraint
-for each formal parameter which is covariant-by-declaration
-as the constraint which is specified for a declaration in $C$
-(\ref{instanceMethods}).%
-}
+It is a compile-time error if a class named $C$ declares
+a member with basename (\ref{classMemberConflicts}) $C$.
+If a generic class named $G$ declares a type variable named $X$,
+it is a compile-time error
+if $X$ is equal to $G$,
+or if $G$ has a member whose basename is $X$,
+or if $G$ has a constructor named \code{$G$.$X$}.
```

Comment:
(We haven't defined the basename of a constructor name *G.X* to be *G* yet?)

Reply:
No, basename still simply strips a trailing `=`.